nbbuild/antsrc/org/netbeans/nbbuild/CustomJavac.java

```diff
@@ -226,12 +226,16 @@ private void cleanUpDependDebris() {
                 continue;
             }
             int i = clazz.indexOf('$');
-            File enclosing = new File(d, clazz.substring(0, i) + ".class");
-            if (!enclosing.isFile()) {
-                File enclosed = new File(d, clazz);
-                log(clazz + " will be deleted since " + enclosing.getName() + " is missing", Project.MSG_VERBOSE);
-                if (!enclosed.delete()) {
-                    throw new BuildException("could not delete " + enclosed, getLocation());
+            // ignore filenames that start right with '$' (separatorChar preceded), these could not be inner classes.
+            if (i > 0 && clazz.charAt(i - 1) != File.separatorChar) {
+                File enclosing = new File(d, clazz.substring(0, i) + ".class");
```

Comment:
clazz.substring(0, i).concat(".class") will avoid one allocation and one loop

Reply:
> clazz.substring(0, i).concat(".class") will avoid one allocation and one loop

I can't see that - could you explain, especially the "loop" part ?

Reply:
Let's have some fun. I was curious, so I wrote a dumb testcase:
```
        long n1 = System.nanoTime();
        for (long i = 0; i < 100000000L; i++) {
            int index = S.indexOf('$');
            String x = S.substring(0, index) + ".class";
            A[0] = x;
        }
        long n2 = System.nanoTime();
        for (long i = 0; i < 100000000L; i++) {
            int index = S.indexOf('$');
            String x = S.substring(0, index).concat(".class");
            A[0] = x;
        }
        long n3 = System.nanoTime();
        
        System.err.println("First: " + (n2 - n1));
        System.err.println("Second: " + (n3 - n2));
```

What was funy was the result:
```
First:      3339608014
Second: 3703034326
```

Reply:
i have also an older JMH benchmark to offer:
```java
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Warmup(iterations = 5, time = 5, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 5, time = 5, timeUnit = TimeUnit.SECONDS)
@Fork(1)
@State(Scope.Benchmark)
public class ConcatJMH {

    int i = 0;

    @Setup(Iteration)
    public void setup() {
        i = 0;
    }

    @Benchmark
    public void chain(Blackhole bh) {
        StringBuilder sb = new StringBuilder();
        sb.append("a");
        sb.append("[").append(i).append("]");
        sb.append("c");
        bh.consume(sb.toString());
    }

    @Benchmark
    public void mix(Blackhole bh) {
        StringBuilder sb = new StringBuilder();
        sb.append("a");
        sb.append("["+i+"]");
        sb.append("c");
        bh.consume(sb.toString());
    }

    @Benchmark
    public void plus(Blackhole bh) {
        bh.consume("a"+"["+i+"]"+"c");
    }

    @Benchmark
    public void concat(Blackhole bh) {
        bh.consume("a".concat("[").concat(String.valueOf(i)).concat("]").concat("c"));
    }

    public static void main(String[] args) throws RunnerException {
        
        Options opt = new OptionsBuilder()
                .include(ConcatJMH.class.getSimpleName())
                .resultFormat(ResultFormatType.JSON)
                .result("results/"+ConcatJMH.class.getSimpleName()+".json")
                .build();

        new Runner(opt).run();

    }
}
```
results on JDK 19:
```
    Benchmark          Mode  Cnt   Score   Error  Units
    ConcatJMH.chain    avgt    5  11.478 ± 0.266  ns/op
    ConcatJMH.mix      avgt    5  19.877 ± 0.301  ns/op
    ConcatJMH.concat   avgt    5   8.388 ± 0.179  ns/op
```

good old '+' wins hands down :)

Reply:
thanks for the feedback, I learned something new today :) 
it's good that the most compact form of concatenation is finally the most performant one
let's hope NetBeans doesn't get stuck with java8 forever ;)